### PR TITLE
Gutenberg: don't crash on unsupported block types, instead replace with unsupported block placeholder block 

### PR DESCRIPTION
**Pull request by iPegii**
Create unsupported block, add handling

### CODE REVIEW

File: services/cms/src/blocks/UnsupportedBlock/index.tsx [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { BlockConfiguration } from "@wordpress/blocks"
+
+import UnsupportedEditor from "./UnsupportedEditor"
+import UnsupportedSave from "./UnsupportedSave"
+
+const UnsupportedConfiguration: BlockConfiguration = {
```
Comment (nygrenh) [inline services/cms/src/blocks/UnsupportedBlock/index.tsx:6]:
```suggestion
const UnsupportedBlock: BlockConfiguration = {
```

File: services/cms/src/blocks/UnsupportedBlock/index.tsx [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  title: "Unsupported block",
+  description: "Unsupported component block.",
+  category: "design",
+  attributes: {},
+  edit: UnsupportedEditor,
+  save: UnsupportedSave,
+}
+
+export default UnsupportedConfiguration
```
Comment (nygrenh) [inline services/cms/src/blocks/UnsupportedBlock/index.tsx:15]:
```suggestion
export default UnsupportedBlock
```

File: services/cms/src/blocks/UnsupportedBlock/UnsupportedEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import { InnerBlocks } from "@wordpress/block-editor"
+import { BlockEditProps } from "@wordpress/blocks"
+import React from "react"
+
+import BlockPlaceholderWrapper from "../BlockPlaceholderWrapper"
+
+const ALLOWED_NESTED_BLOCKS = [""]
```
Comment (nygrenh) [inline services/cms/src/blocks/UnsupportedBlock/UnsupportedEditor.tsx:7]:
```suggestion
```

File: services/cms/src/blocks/UnsupportedBlock/UnsupportedEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+const UnsupportedEditor: React.FC<BlockEditProps<Record<string, never>>> = ({ clientId }) => {
+  return (
+    <BlockPlaceholderWrapper id={clientId}>
+      <h3>Unsupported block placeholder</h3>
+      <p>This block is used as placeholder for blocks that are unsupported.</p>
+      <InnerBlocks allowedBlocks={ALLOWED_NESTED_BLOCKS} />
```
Comment (nygrenh) [inline services/cms/src/blocks/UnsupportedBlock/UnsupportedEditor.tsx:14]:
```suggestion
```

File: services/cms/src/components/editors/PageEditor.tsx [hunk 5]
```diff
@@ -38,7 +38,15 @@ const PageEditor: React.FC<PageEditorProps> = ({ data, handleSave }) => {
       chapter_id: data.chapter_id,
       front_page_of_chapter_id: null,
     })
-    setContent(res.content as BlockInstance[])
+    const resContent: BlockInstance[] = res.content as BlockInstance[]
```
Comment (nygrenh) [inline services/cms/src/components/editors/PageEditor.tsx:41]: before handlesave

File: services/cms/src/components/editors/GutenbergEditor.tsx [hunk 6]
```diff
@@ -70,8 +71,23 @@ const GutenbergEditor: React.FC<GutenbergEditorProps> = ({
   }, [courseId])
 
   const handleChanges = (newContent: BlockInstance[]): void => {
-    console.log(newContent)
-    onContentChange(newContent)
+    const modfiedNewContent = newContent.map((block) => {
```
Comment (nygrenh) [inline services/cms/src/components/editors/GutenbergEditor.tsx:74]: do this in this component/in the parent of this component before we pass the content to BlockEditorProvider

**Review by nygrenh — CHANGES_REQUESTED** [review]

File: services/cms/src/components/editors/PageEditor.tsx [hunk 7]
```diff
@@ -29,26 +29,54 @@ const PageEditor: React.FC<PageEditorProps> = ({ data, handleSave }) => {
...
     })
-    const resContent: BlockInstance[] = res.content as BlockInstance[]
-    const modfiedResContent = resContent.map((block) => {
-      if (block.name === "moocfi/unsupported-block-type") {
-        return block.attributes.originalBlockJson
+    setContent(res.content as BlockInstance[])
+    setSaving(false)
+  }
+
+  useEffect(() => {
```
Comment (nygrenh) [inline services/cms/src/components/editors/PageEditor.tsx:57]: https://reactjs.org/docs/hooks-reference.html#usememo

File: services/cms/src/components/editors/PageEditor.tsx [hunk 8]
```diff
@@ -29,26 +29,54 @@ const PageEditor: React.FC<PageEditorProps> = ({ data, handleSave }) => {
...
+    const allSupportedBlocks = supportedCoreBlocks.concat(
```
Comment (nygrenh) [inline services/cms/src/components/editors/PageEditor.tsx:58]: check if top level page from page.chapter_id and select which one to use

**Review by nygrenh — CHANGES_REQUESTED** [review]

File: services/cms/src/components/editors/PageEditor.tsx [hunk 9]
```diff
@@ -54,13 +54,13 @@ const PageEditor: React.FC<PageEditorProps> = ({ data, handleSave }) => {
     setSaving(false)
   }
 
-  useEffect(() => {
-    const allSupportedBlocks = supportedCoreBlocks.concat(
-      supportedBlocksForPages,
-      supportedBlocksTopLevelPages,
-    )
+  const allSupportedBlocks = urlPath.startsWith("/chapter-1/")
```
Comment (nygrenh) [inline services/cms/src/components/editors/PageEditor.tsx:57]:
```suggestion
  const allSupportedBlocks = data.chapter_id
```

**Review by nygrenh — CHANGES_REQUESTED** [review]
Also implement in email editor

**Review by nygrenh — CHANGES_REQUESTED** [review]

File: services/cms/src/utils/Gutenberg/UseBlocksWithUnsupportedBlocksRemoved.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
+import { BlockInstance } from "@wordpress/blocks"
+
+export function UseBlocksWithUnsupportedBlocksRemoved(
```
Comment (nygrenh) [inline services/cms/src/utils/Gutenberg/UseBlocksWithUnsupportedBlocksRemoved.ts:3]:
```suggestion
export function useBlocksWithUnsupportedBlocksRemoved(
```
Also rename the file

File: services/cms/src/components/editors/PageEditor.tsx [hunk 11]
```diff
@@ -59,24 +54,8 @@ const PageEditor: React.FC<PageEditorProps> = ({ data, handleSave }) => {
     : supportedCoreBlocks.concat(supportedBlocksTopLevelPages)
 
   useMemo(() => {
```
Comment (nygrenh) [inline services/cms/src/components/editors/PageEditor.tsx:56]: Move useMemo to the hook.

```js
    const contentWIthUsupportedBlocksRemoved =  useBlocksWithUnsupportedBlocksRemoved(content, allSupportedBlocks)
```

File: services/cms/src/utils/Gutenberg/UseBlocksWithUnsupportedBlocksRemoved.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  blocks: BlockInstance[],
+  supportedBlocks: string[],
+): BlockInstance[] {
+  const modifiedBlocks = blocks.map((block) => {
```
Comment (nygrenh) [inline services/cms/src/utils/Gutenberg/UseBlocksWithUnsupportedBlocksRemoved.ts:7]: react.memo

**Review by nygrenh — CHANGES_REQUESTED** [review]

File: services/cms/src/utils/Gutenberg/useBlocksWithUnsupportedBlocksRemoved.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    } else {
+      return block
+    }
+  })
+  return modifiedBlocks
+}
+
+/*-----
+
+export default function useBlocksWithUnsupportedBlocksRemoved(
```
Comment (nygrenh) [inline services/cms/src/utils/Gutenberg/useBlocksWithUnsupportedBlocksRemoved.ts:34]: remove the dead code

**Review by nygrenh — CHANGES_REQUESTED** [review]

File: services/cms/src/components/editors/PageEditor.tsx [hunk 14]
```diff
@@ -32,21 +34,34 @@ const PageEditor: React.FC<PageEditorProps> = ({ data, handleSave }) => {
   const [content, setContent] = useState<BlockInstance[]>(data.content as BlockInstance[])
```
Comment (nygrenh) [inline services/cms/src/components/editors/PageEditor.tsx:34]: const [content, setContent] = useState<BlockInstance[]>(modifyBlocks(data.content) as BlockInstance[])

**Review by nygrenh — CHANGES_REQUESTED** [review]

File: services/cms/src/utils/Gutenberg/modifyBlocks.ts [hunk 15]
```diff
@@ -1,5 +1,5 @@
 import { BlockInstance } from "@wordpress/blocks"
-import { useMemo } from "react"
+/*import { useMemo } from "react"
```
Comment (nygrenh) [inline services/cms/src/utils/Gutenberg/modifyBlocks.ts:2]: dead code

**Review by nygrenh — CHANGES_REQUESTED** [review]

**Review by sebazai — APPROVED** [review]

**Pull request merged**